File: react-native/storyly-placement-react-native/ios/Common/Data/Product/ProductData.swift
import Foundation
import StorylyPlacement
import StorylyCore

// MARK: - STRProductItem Encoding/Decoding

func encodeSTRProductItem(_ item: STRProductItem) -> [String: Any] {
    let result: [String: Any] = ([
        "productId": item.productId,
        "productGroupId": item.productGroupId,
        "title": item.title,
        "url": item.url,
        "desc": item.desc,
        "price": item.price,
        "currency": item.currency,
        "salesPrice": item.salesPrice,
        "lowestPrice": item.lowestPrice,
        "imageUrls": item.imageUrls,
        "ctaText": item.ctaText,
        "variants": item.variants?.map { encodeSTRProductVariant($0) }
    ] as [String: Any?]).compactMapValues { $0 }
    return result
}

func decodeSTRProductItem(_ dict: [String: Any]) -> STRProductItem? {
    guard let productId = dict["productId"] as? String,
          let productGroupId = dict["productGroupId"] as? String,
          let title = dict["title"] as? String,
          let url = dict["url"] as? String,
          let desc = dict["desc"] as? String,
          let currency = dict["currency"] as? String,
          let price = dict["price"] as? Double else {
        return nil
    }
    
    
    let salesPrice = dict["salesPrice"] as? Double
    let imageUrls = dict["imageUrls"] as? [String]
    let ctaText = dict["ctaText"] as? String
    let variants: [STRProductVariant]? = (dict["variants"] as? [[String: Any]])?.compactMap { decodeSTRProductVariant($0) }
    let lowestPrice = dict["lowestPrice"] as? NSNumber
    
    return STRProductItem(
        productId: productId,
        productGroupId: productGroupId,
        title: title,
        url: url,
        description: desc,
        price: Float(price),
        salesPrice: salesPrice as NSNumber?,
        lowestPrice: lowestPrice,
        currency: currency,
        imageUrls: imageUrls,
        variants: variants,
        ctaText: ctaText
    )
}

// MARK: - STRProductVariant Encoding/Decoding

func encodeSTRProductVariant(_ variant: STRProductVariant) -> [String: Any] {
    let result: [String: Any] = [
        "name": variant.name,
        "value": variant.value,
        "key": variant.key
    ]
    return result
}

func decodeSTRProductVariant(_ dict: [String: Any]) -> STRProductVariant? {
    guard let name = dict["name"] as? String,
          let value = dict["value"] as? String,
          let key = dict["key"] as? String else {
        return nil
    }
    return STRProductVariant(name: name, value: value, key: key)
}

// MARK: - STRCart Encoding/Decoding

func encodeSTRCart(_ cart: STRCart?) -> [String: Any]? {
    guard let cart = cart else { return nil }
    
    let result: [String: Any] = ([
        "items": cart.items.map { encodeSTRCartItem($0) },
        "totalPrice": cart.totalPrice,
        "currency": cart.currency,
        "oldTotalPrice": cart.oldTotalPrice
    ] as [String: Any?]).compactMapValues { $0 }
    
    return result
}

func decodeSTRCart(_ dict: [String: Any]?) -> STRCart? {
    guard let dict = dict,
          let itemsArray = dict["items"] as? [[String: Any]],
          let totalPrice = dict["totalPrice"] as? Double,
          let currency = dict["currency"] as? String else {
        return nil
    }
    let oldTotalPrice = dict["oldTotalPrice"] as? Double
    return STRCart(
        items: itemsArray.compactMap { decodeSTRCartItem($0) },
        totalPrice: Float(totalPrice),
        oldTotalPrice: oldTotalPrice as NSNumber?,
        currency: currency
    )
}

// MARK: - STRCartItem Encoding/Decoding

func encodeSTRCartItem(_ item: STRCartItem?) -> [String: Any?]? {
    guard let item = item else { return nil }
    
    let result: [String: Any] = ([
        "item": encodeSTRProductItem(item.item),
        "quantity": item.quantity,
        "totalPrice": item.totalPrice,
        "oldTotalPrice": item.oldTotalPrice,
    ] as [String: Any?]).compactMapValues { $0 }
    return result
}

func decodeSTRCartItem(_ dict: [String: Any]) -> STRCartItem? {
    guard let itemDict = dict["item"] as? [String: Any],
          let item = decodeSTRProductItem(itemDict),
          let quantity = dict["quantity"] as? Int else {
        return nil
    }
    
    let totalPrice = dict["totalPrice"] as? Double
    let oldTotalPrice = dict["oldTotalPrice"] as? Double
    
    return STRCartItem(
        item: item,
        quantity: quantity,
        totalPrice: totalPrice as NSNumber?,
        oldTotalPrice: oldTotalPrice as NSNumber?
    )
}

// MARK: - STRProductInformation Encoding

func encodeSTRProductInformation(_ info: STRProductInformation) -> [String: Any?] {
    return [
        "productId": info.productId,
        "productGroupId": info.productGroupId
    ]
}

File: react-native/storyly-placement-react-native/android/src/main/java/com/storylyplacementreactnative/common/data/product/ProductData.kt
package com.storylyplacementreactnative.common.data.product

import com.appsamurai.storyly.core.data.model.product.STRCart
import com.appsamurai.storyly.core.data.model.product.STRCartItem
import com.appsamurai.storyly.core.data.model.product.STRProductInformation
import com.appsamurai.storyly.core.data.model.product.STRProductItem
import com.appsamurai.storyly.core.data.model.product.STRProductVariant

internal fun encodeSTRProductItem(product: STRProductItem) {
    mapOf(
        "productId" to product.productId,
        "productGroupId" to product.productGroupId,
        "title" to product.title,
        "url" to product.url,
        "desc" to product.desc,
        "price" to product.price.toDouble(),
        "salesPrice" to product.salesPrice?.toDouble(),
        "lowestPrice" to product.lowestPrice?.toDouble(),
        "currency" to product.currency,
        "imageUrls" to product.imageUrls,
        "variants" to product.variants.map { variant -> encodeSTRProductVariant(variant) },
        "ctaText" to product.ctaText,
    )
}

internal fun decodeSTRProductItem(product: Map<String, Any?>?): STRProductItem? {
    product ?: return null
    val productId = product["productId"] as? String ?: return null
    val productGroupId = product["productGroupId"] as? String ?: return null
    val title = product["title"] as? String ?: return null
    val url = product["url"] as? String ?: return null
    val desc = product["desc"] as? String ?: return null
    val price = (product["price"] as? Double)?.toFloat() ?: return null
    return STRProductItem(
        productId = productId,
        productGroupId = productGroupId,
        title = title,
        url = url,
        desc = desc,
        price = price,
        salesPrice = (product["salesPrice"] as? Double)?.toFloat(),
        lowestPrice = (product["lowestPrice"] as? Double)?.toFloat(),
        currency = product["currency"] as String,
        imageUrls = product["imageUrls"] as? List<String>,
        variants = (product["variants"] as? List<Map<String, Any?>>)
            ?.mapNotNull { variant -> decodeSTRProductVariant(variant) } ?: emptyList(),
        ctaText = product["ctaText"] as? String,
    )
}

internal fun encodeSTRProductVariant(variant: STRProductVariant): Map<String, Any?> {
    return mapOf(
        "name" to variant.name,
        "value" to variant.value,
        "key" to variant.key,
    )
}

internal fun decodeSTRProductVariant(variant: Map<String, Any?>): STRProductVariant? {
    val name = variant["name"] as? String ?: return null
    val value = variant["value"] as? String ?: return null
    val key = variant["key"] as? String ?: return null
    return STRProductVariant(
        name = name,
        value = value,
        key = key,
    )
}

internal fun encodeSTRProductInformation(info: STRProductInformation): Map<String, Any?> {
    return mapOf(
        "productId" to info.productId,
        "productGroupId" to info.productGroupId,
    )
}

internal fun encodeSTRCart(cart: STRCart?): Map<String, Any?>? {
    cart ?: return null
    return mapOf(
     "items" to cart.items.map { item -> encodeSTRCartItem(item) },
     "totalPrice" to cart.totalPrice,
     "oldTotalPrice" to cart.oldTotalPrice,
     "currency" to cart.currency,
    )
}

internal fun decodeSTRCart(cart: Map<String, Any?>?): STRCart? {
    cart ?: return null
    return STRCart(
        items = (cart["items"] as? List<Map<String, Any?>>)
            ?.mapNotNull { item -> decodeSTRCartItem(item) }
            ?: emptyList(),
        totalPrice = (cart["totalPrice"] as Double).toFloat(),
        oldTotalPrice = (cart["oldTotalPrice"] as? Double)?.toFloat(),
        currency = cart["currency"] as String,
    )
}

internal fun encodeSTRCartItem(item: STRCartItem?): Map<String, Any?>? {
    item ?: return null
    return mapOf(
        "item" to encodeSTRProductItem(item.item),
        "quantity" to item.quantity,
        "totalPrice" to item.totalPrice,
        "oldTotalPrice" to item.oldTotalPrice,
    )
}

internal fun decodeSTRCartItem(item: Map<String, Any?>?): STRCartItem? {
    item ?: return null
    val productItem = item["item"] as? Map<String, Any?> ?: return null
    val cartItem = decodeSTRProductItem(productItem) ?: return null
    val quantity = item["quantity"] as? Int ?: return null
    return STRCartItem(
        item = cartItem,
        quantity = quantity,
        totalPrice = (item["totalPrice"] as? Double)?.toFloat(),
        oldTotalPrice = (item["oldTotalPrice"] as? Double)?.toFloat(),
    )
}
